File: src/minecraft-server.js
const child_process = require('child_process');

module.exports = {

    MinecraftServer : class MinecraftServer {

        constructor(port, command) {
            this.port = port;
            this.command = command;
            this.process = null;
            this.startCallbacks = [];
            this.readyCallbacks = [];
            this.exitCallbacks = [];
        }
    
        start() {

            if (this.process != null) {throw new Error("Minecraft server has already started.");}

            const a = this.command.split(/\s+/);
            const c = a.shift();
            this.process = child_process.spawn(c, a);
            this.startCallbacks.forEach(callback => callback(this));
            this.process.on('exit', (c, s) => {this.exitCallbacks.forEach(callback => callback(this)); this.process = null;});

            this.process.stdout.on('data', (data) => {
                
                const output = data.toString();
                const lines = output.split('\n');
                const pattern = /^\[\d{2}:\d{2}:\d{2}\] \[Server thread\/INFO\]: Done.*/;
                
                for (let i = 0; i < lines.length; i++) {
                    const line = lines[i];
                    if (pattern.test(line)) {
                        this.readyCallbacks.forEach(callback => callback(this));
                        break;
                    }   
                }

            });

        }
    
        on(state, callback) {
            if (state.toLowerCase() == 'start') {this.startCallbacks.push(callback);}
            if (state.toLowerCase() == 'ready') {this.readyCallbacks.push(callback);}
            if (state.toLowerCase() == 'exit') {this.exitCallbacks.push(callback);}
        }

        stop() {
            if (this.process != null) {
                this.process.stdin.write('stop\n'); 
                this.process.stdin.end();
            }
        }
    
        destroy() {
            if (this.process != null) {this.process.kill();}
        }
    
    }    

}

File: scripts/minecraft_server.py
import subprocess
import threading
import re

class MinecraftServer:

    def __init__(self, command):
        
        self.command = command
        self.callbacks = {'start': [], 'ready': [], 'exit': []}
        self.process = None

    def __del__(self):
        if self.process != None: self.process.terminate()

    def start(self):

        if self.process != None: raise Exception("Minecraft server has already started.")

        def routine():

            self.process = subprocess.Popen(["java", "-Xms1G", "-Xmx1G", "-jar", "server.jar", "nogui"], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
            self._run_event('start')

            def waiter(p):
                p.wait()
                self._run_event('exit')
                if p == self.process: self.process = None

            thread = threading.Thread(target=waiter, args=(self.process,))
            thread.start()

            while self.process != None:
                
                line = self.process.stdout.readline()
                if not line: break
                print(line, end="")

                pattern = re.compile(r"^\[\d{2}:\d{2}:\d{2}\] \[Server thread/INFO\]: Done.*")
                result = pattern.match(line)
                if result: self._run_event('ready')

        thread = threading.Thread(target=routine, args=())
        thread.start()

    def stop(self):
        if self.process == None: return
        self.process.stdin.write('stop\n')
        self.process.stdin.flush()

    def enter_command(self, command):
        if self.process == None: return
        self.process.stdin.write(command)
        self.process.stdin.flush()

    def _run_event(self, event):
        if event not in self.callbacks: return
        for callback in self.callbacks[event]: callback()

    def on(self, event, callback):
        if event not in self.callbacks: return
        self.callbacks[event].append(callback)
